tests/test_z3_visitor.py

```diff
@@ -41,16 +30,127 @@ def f(x: bool, y: bool, z: bool, a):
+        '''
+        pass
+    """,
+    """
+    def in_set(x: int):
+        '''
+        Preconditions:
+            - x in {1, 2, 3}
+        '''
+        pass
+    """,
+    """
+    def in_tuple(x: int):
+        '''
+        Preconditions:
+            - x in (1, 2, 3)
+        '''
+        pass
+    """,
+    """
+    def not_in_list(x: int):
+        '''
+        Preconditions:
+            - x not in [1, 2, 3]
+        '''
+        pass
+    """,
+    """
+    def not_in_set(x: int):
+        '''
+        Preconditions:
+            - x not in {1, 2, 3}
+        '''
+        pass
+    """,
+    """
+    def not_in_tuple(x: int):
+        '''
+        Preconditions:
+            - x not in (1, 2, 3)
+        '''
+        pass
+    """,
+    """
+    def in_empty_container(x: int):
+        '''
+        Preconditions:
+            - x in []
+            - x in set()
+            - x in ()
+        '''
+        pass
+    """,
+    """
+    def not_in_empty_container(x: int):
+        '''
+        Preconditions:
+            - x not in []
+            - x not in set()
+            - x not in ()
+        '''
+        pass
+    """,
+]
+
+
+# expected arithmetic expressions
+x = z3.Int("x")
+y = z3.Int("y")
+z = z3.Real("z")
+arithmetic_expected = [[x**2 + y**2 == z**2, z3.And([x > 0, y > 0, z == 0])]]
+
+# expected boolean expressions
+x = z3.Bool("x")
+y = z3.Bool("y")
+z = z3.Bool("z")
+boolean_expected = [[z3.And([x, y, z]), z3.Not(z3.Or([x, y, z]))]]
+
+# expected container expressions
+x = z3.Int("x")
+container_expected = [
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.BoolVal(False), z3.BoolVal(False)],
+    [z3.BoolVal(True), z3.BoolVal(True)],
+]
+
+# lists of all test cases
+code_list = [arithmetic_list, boolean_list, container_list]
+expected_list = [arithmetic_expected, boolean_expected, container_expected]
+
+
+def _get_constraints_from_code(code) -> List[z3.ExprRef]:
+    """
+    Return the z3 constraints of the given function
+    """
+    z3v = Z3Visitor()
     mod = z3v.visitor.visit(astroid.parse(code))
     function_def = mod.body[0]
-    # Declare variables
-    x = z3.Bool("x")
-    y = z3.Bool("y")
-    z = z3.Bool("z")
-    # Construct expected
-    expected = [z3.And([x, y, z]), z3.Not(z3.Or([x, y, z]))]
-    actual = function_def.z3_constraints
-    for e, a in zip(expected, actual):
-        solver = z3.Solver()
-        solver.add(e == a)
-        assert solver.check() == z3.sat
+    return function_def.z3_constraints
+
+
+@pytest.mark.parametrize("code, expected", zip(code_list, expected_list))
+def test_constraint(code, expected):
+    for function, function_expected in zip(code, expected):
+        function_actual = _get_constraints_from_code(function)
+        assert len(function_actual) == len(function_expected)
```

Comment:
Although checking the length of expected and actual expressions to be equal has been our existing approach, and is used to ensure `zip` function works correctly, I'm not sure this is the best practice to perform the test. In principle, we do not need the number of expressions to be equal to that asseted, or the expressions appear in the exact same order, as long as the two list of expressions are equivalent. Our approach seems to be testing against our specifc implementation instead of the module's feature.

Reply:
While the relevant module functionality is to convert the preconditions into Z3 expressions, it is appropriate for the module to do this conversion as a one-to-one mapping (i.e., without further simplification) until we decide to change this.

More to the point, however, this reveals a limitation on our current testing approach. The changes you made were to `ExprWrapper`, which currently is only tested indirectly through the tests in this file for `Z3Visitor`. Certainly I would not accept any simplification in this file until there were unit tests for `ExprWrapper` itself. (I'm not saying to add those in this PR, but rather to explain why _even if I approve simplification of the precondition expressions_ generally, I would not approve changing the tests in this file without additional tests added.)